### PR TITLE
feat(flow): transform substrait SELECT&WHERE&GROUP BY to Flow Plan

### PR DESCRIPTION
**Pull request by discord9**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## What's changed and what's your intention?

Added some basic transform functions to transform from Substrait Plan to Flow Plan,
and add some testcase to cover most case currently impled(Please help me found more useful testcase)

- support transform simple Substrait Plan including `SELECT`, `WHERE` and `GROUP BY`
- support  simple binary operators like `1+number*2/3%4>0` and simple constant folding
- related tests are added(more tests are welcomed)

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.
- [x]  This PR does not require documentation updates.

### CODE REVIEW

File: src/flow/src/transform.rs [hunk 1]
```diff
@@ -0,0 +1,1221 @@
...
+    // Parse relations
+    match plan.relations.len() {
+        1 => {
+            match plan.relations[0].rel_type.as_ref() {
+                Some(rt) => match rt {
+                    plan_rel::RelType::Rel(rel) => {
+                        Ok(from_substrait_rel(ctx, rel, &function_extension)?)
+                    },
+                    plan_rel::RelType::Root(root) => {
+                        Ok(from_substrait_rel(ctx, root.input.as_ref().unwrap(), &function_extension)?)
```
Comment (zhongzc) [inline src/flow/src/transform.rs:142]: Is it safe to unwrap?

File: src/flow/src/transform.rs [hunk 2]
```diff
@@ -0,0 +1,1221 @@
...
+        .extensions
+        .iter()
+        .map(|e| match &e.mapping_type {
+            Some(ext) => match ext {
+                MappingType::ExtensionFunction(ext_f) => Ok((ext_f.function_anchor, &ext_f.name)),
+                _ => not_impl_err!("Extension type not supported: {ext:?}"),
+            },
+            None => not_impl_err!("Cannot parse empty extension"),
+        })
+        .collect::<Result<HashMap<_, _>, Error>>()?;
```
Comment (zhongzc) [inline src/flow/src/transform.rs:131]: It seems that extensions play a crucial role. Can they be extracted into a function and explained in detail in the doc comment (in a way that is friendly to beginners like me)?

File: src/flow/src/transform.rs (hunk 1 above)
Comment (discord9) [reply src/flow/src/transform.rs:142]: > Is it safe to unwrap?

Fixed a bunch of unwrap

File: src/flow/src/transform/plan.rs [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+                            let input_arity = get_table.typ.column_types.len();
+                            let mfp = MapFilterProject::new(input_arity)
+                                .project(column_indices.clone())?;
+                            get_table.mfp(mfp)
+                        }
+                        _ => Ok(get_table),
+                    }
+                }
+                _ => not_impl_err!("Only NamedTable reads are supported"),
+            },
```
Comment (zhongzc) [inline src/flow/src/transform/plan.rs:146]:
```suggestion
            Some(RelType::Read(read)) => {
                if let Some(ReadType::NamedTable(nt)) = &read.read_type {
                    let (table_id, table_type) = ctx.table(&nt.names)?;
                    let get_table_plan = Plan::Get {
                        id: crate::expr::Id::Global(table_id),
                    };
                    let plan = TypedPlan {
                        typ: table_type,
                        plan: get_table_plan,
                    };

                    if let Some(MaskExpression {
                        select: Some(projection),
                        ..
                    }) = &read.projection
                    {
                        let column_indices = projection
                            .struct_items
                            .iter()
                            .map(|item| item.field as usize)
                            .collect_vec();
                        let input_arity = plan.typ.column_types.len();
                        let mfp = MapFilterProject::new(input_arity).project(column_indices)?;
                        plan.mfp(mfp)
                    } else {
                        Ok(plan)
                    }
                } else {
                    not_impl_err!("Only NamedTable reads are supported")
                }
            }
```

File: src/flow/src/transform/expr.rs [hunk 4]
```diff
@@ -0,0 +1,448 @@
...
+use crate::transform::FunctionExtensions;
+
+impl TypedExpr {
+    /// Convert ScalarFunction into Flow's ScalarExpr
+    pub fn from_substrait_scalar_func(
+        f: &ScalarFunction,
+        input_schema: &RelationType,
+        extensions: &FunctionExtensions,
+    ) -> Result<TypedExpr, Error> {
+        let fn_name = extensions.get(&f.function_reference).ok_or_else(|| {
```
Comment (zhongzc) [inline src/flow/src/transform/expr.rs:43]: `with_context` instead

File: src/flow/src/transform/expr.rs [hunk 5]
```diff
@@ -0,0 +1,448 @@
...
+                        then: Box::new(then.expr),
+                        els: Box::new(build_if_then(next_if_then, els).expr),
+                    },
+                    then.typ,
+                )
+            } else {
+                els
+            }
+        }
+        let expr_if = build_if_then(ifs.into_iter(), els);
```
Comment (zhongzc) [inline src/flow/src/transform/expr.rs:205]:
```suggestion

        fn build_if_then_recur(
            mut next_if_then: impl Iterator<Item = (TypedExpr, TypedExpr)>,
            els: TypedExpr,
        ) -> TypedExpr {
            if let Some((cond, then)) = next_if_then.next() {
                // always assume the type of `if` expr is the same with the `then` expr
                TypedExpr::new(
                    ScalarExpr::If {
                        cond: Box::new(cond.expr),
                        then: Box::new(then.expr),
                        els: Box::new(build_if_then_recur(next_if_then, els).expr),
                    },
                    then.typ,
                )
            } else {
                els
            }
        }
        let expr_if = build_if_then_recur(ifs.into_iter(), els);
```

File: src/flow/src/transform/aggr.rs [hunk 6]
```diff
@@ -0,0 +1,445 @@
...
+        extensions: &FunctionExtensions,
+    ) -> Result<Vec<AggregateExpr>, Error> {
+        let _ = ctx;
+        let mut aggr_exprs = vec![];
+
+        for m in measures {
+            let filter = match &m.filter {
+                Some(fil) => Some(TypedExpr::from_substrait_rex(fil, typ, extensions)?),
+                None => None,
+            };
```
Comment (zhongzc) [inline src/flow/src/transform/aggr.rs:99]:
```suggestion
            let filter = m
                .filter
                .as_ref()
                .map(|fil| TypedExpr::from_substrait_rex(fil, typ, extensions))
                .transpose()?;
```

File: src/flow/src/transform/aggr.rs [hunk 7]
```diff
@@ -0,0 +1,445 @@
...
+                        _ => false,
+                    };
+                    AggregateExpr::from_substrait_agg_func(
+                        f, typ, extensions, filter, // TODO(discord9): impl order_by
+                        None, distinct,
+                    )
+                }
+                None => not_impl_err!("Aggregate without aggregate function is not supported"),
+            };
+            aggr_exprs.push(agg_func?.clone());
```
Comment (zhongzc) [inline src/flow/src/transform/aggr.rs:115]:
```suggestion
            let agg_func = match &m.measure {
                Some(f) => {
                    let distinct = match f.invocation {
                        _ if f.invocation == AggregationInvocation::Distinct as i32 => true,
                        _ if f.invocation == AggregationInvocation::All as i32 => false,
                        _ => false,
                    };
                    AggregateExpr::from_substrait_agg_func(
                        f, typ, extensions, filter, // TODO(discord9): impl order_by
                        None, distinct,
                    )?
                }
                None => {
                    return not_impl_err!("Aggregate without aggregate function is not supported")
                }
            };
            aggr_exprs.push(agg_func);
```

File: src/flow/src/transform/aggr.rs [hunk 8]
```diff
@@ -0,0 +1,445 @@
...
+    fn from_substrait_gen_key_val_plan(
+        aggr_exprs: &mut [AggregateExpr],
+        group_exprs: &[TypedExpr],
+        input_arity: usize,
+    ) -> Result<KeyValPlan, Error> {
+        let (group_expr_val, _group_expr_type): (Vec<_>, Vec<_>) = group_exprs
+            .iter()
+            .cloned()
+            .map(|TypedExpr { expr, typ }| (expr, typ))
+            .unzip();
```
Comment (zhongzc) [inline src/flow/src/transform/aggr.rs:179]:
```suggestion
        let group_expr_val = group_exprs
            .iter()
            .map(|expr| expr.expr.clone())
            .collect_vec();
```

File: src/flow/src/transform/aggr.rs [hunk 9]
```diff
@@ -0,0 +1,445 @@
...
+                    .map(|(idx, aggr)| {
+                        let ret = aggr.expr.clone();
+                        aggr.expr = ScalarExpr::Column(idx);
+                        ret
+                    })
+                    .collect_vec();
+                let aggr_arity = aggr_exprs.len();
+
+                MapFilterProject::new(input_arity)
+                    .map(input_exprs.clone())?
```
Comment (zhongzc) [inline src/flow/src/transform/aggr.rs:203]:
```suggestion
                    .map(input_exprs)?
```

File: src/flow/src/transform.rs [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
+        }.fail()
+    };
+}
+
+mod aggr;
+mod expr;
+mod literal;
+mod plan;
+
+use literal::{from_substrait_literal, from_substrait_type};
```
Comment (zhongzc) [inline src/flow/src/transform.rs:83]:
```suggestion
mod aggr;
mod expr;
mod literal;
mod plan;

use std::collections::HashMap;

use snafu::OptionExt;
use substrait::substrait_proto::proto::extensions::simple_extension_declaration::MappingType;
use substrait::substrait_proto::proto::extensions::SimpleExtensionDeclaration;

use crate::adapter::error::{Error, NotImplementedSnafu, TableNotFoundSnafu};
use crate::expr::GlobalId;
use crate::repr::RelationType;

/// a simple macro to generate a not implemented error
#[macro_export]
macro_rules! not_impl_err {
    ($($arg:tt)*)  => {
        NotImplementedSnafu {
            reason: format!($($arg)*),
        }.fail()
    };
}

/// generate a plan error
#[macro_export]
macro_rules! plan_err {
    ($($arg:tt)*)  => {
        PlanSnafu {
            reason: format!($($arg)*),
        }.fail()
    };
}
```

**Review by zhongzc — APPROVED** [review]

File: src/flow/src/expr/scalar.rs [hunk 11]
```diff
@@ -24,6 +24,22 @@ use snafu::ensure;
...
 use crate::expr::func::{BinaryFunc, UnaryFunc, UnmaterializableFunc, VariadicFunc};
+use crate::repr::ColumnType;
+
+/// A scalar expression with a known type.
+#[derive(Debug, Clone)]
+pub struct TypedExpr {
+    /// The expression.
+    pub expr: ScalarExpr,
+    /// The type of the expression.
+    pub typ: ColumnType,
```
Comment (fengjiachun) [inline src/flow/src/expr/scalar.rs:35]: Why not `type`? 🤔

**Review by fengjiachun — APPROVED** [review]

File: src/flow/src/expr/scalar.rs (hunk 11 above)
Comment (tisonkun) [reply src/flow/src/expr/scalar.rs:35]: `type` is a keyword, although we can use `r#type`

Comment (tisonkun) [reply src/flow/src/expr/scalar.rs:35]: While it's a common preference debt between `ty` and `typ`. Golang's internal code use `typ` while Rust's internal code use `ty` more commonly.

**Pull request merged**